### ISSUE TITLE
Question: threads an egui

### ISSUE THREAD

**Issue by ckaran** [emilk/egui#110]
How does egui handle threads?  Some GUIs require all interactions be from the main thread, others will allow a secondary thread be designated as the GUI handler (but won't let you switch later on), and still others don't care what thread is making the updates.

What model does egui expect/support?  I want to make sure I never code myself into a corner.

**Comment by emilk** [comment]
Egui is completely agnostic to threads - call it from whatever thread you like!

In the rare case you would want to use the same `egui::Context` from multiple threads at the same time (e.g. to layout different parts of your GUI in parallel) you need to enable the `multi_threaded` feature (or you will get runtime panics). I doubt anyone will actually want to do this, but the possibility exists.

**Comment by ckaran** [comment]
> In the rare case you would want to use the same `egui::Context` from multiple threads at the same time (e.g. to layout different parts of your GUI in parallel) you need to enable the `multi_threaded` feature (or you will get runtime panics). I doubt anyone will actually want to do this, but the possibility exists.

Actually... this is probably going to become more common than you think in the future.  I'm using [`futures`](https://crates.io/crates/futures) in my code (a large scale simulator), and [`executors`](https://crates.io/crates/executors) (which can execute futures on a thread pool) to do the actual execution.  Due to how I chose to set up my code, it would be **MUCH** easier to update the GUI from individual futures than to try to funnel everything into a single thread of control, so I'm grateful that egui is thread agnostic!

**Comment by emilk** [comment]
Drawing the different panels in your screenshot in different threads will be difficult, since egui must know the size of the menu before it can know where to place the toolbar, and needs to know the size of both the menu and toolbar to know where to place the side panels.

In theory this could be helped by fixed sized panels, but how egui is currently written does not permit it within a single egui context (*).

You _should_ be able to run the two side panels in separate threads though (after adding the top panels), but you may run into other problems.

I'd suggest having all egui code in the same thread until you have a performance bottleneck, and when you do: optimize the bottleneck :)

If running different panels and windows in separate threads is still desired, then we need to think through how to let egui support that.

---

(*) You can use different egui contexts for all your panels, and set their `RawInput::screen_rect` appropriately (e.g. with fixed sized offsets from the top). This might be the simplest solution?

**Comment by ckaran** [comment]
It's been a long time since I've looked at the code, so please forgive me if this should be obvious; is the ownership model of egui a true tree? If so, is it possible to render the elements in a bottom-up approach?  My rough thought is that every node in the tree is rendered by a Future.  Each parent Future depends ([joins](https://docs.rs/futures/0.3.17/futures/future/fn.join_all.html)) all of its leaves, which ensures that they will be completed before the parent can be rendered.  As long as the leaves don't depend on sibling leaves, this will work; if there are cases where siblings depend on each other, the parent nodes will need to have the ability to rerun the leaves with appropriate updates (first run might have arguments set to `None`, and second run be set to `Some`).  Not perfect, but it reduces the amount of code that needs to be rerun to fix up stuff that can't be known until after it's been run at least once.

**Comment by emilk** [comment]
There is no ownership model in egui, as there is nothing to own! egui is immediate: https://docs.rs/egui/0.14.2/egui/#understanding-immediate-mode

**Comment by ckaran** [comment]
I think I was being a little unclear; I don't mean that any GUI element lasts for more than one cycle, I meant that the containers are the interior nodes of a tree.  Kind of like this:

```text
GUI
└── Window
    ├── bottom_panel
    │   └── horizontal_1
    │       ├── tool_1
    │       ├── tool_2
    │       └── tool_3
    ├── center_panel
    └── top_panel
        └── horizontal_2
            ├── icon_1
            ├── icon_2
            └── icon_3
```
I'm going to use [`spawn_ok()`](https://docs.rs/futures/0.3.17/futures/executor/struct.ThreadPool.html#method.spawn_ok), [`join!()`](https://docs.rs/futures/0.3.17/futures/macro.join.html), and assume that there exists an async `render` function that is able to render stuff.  Rendering the above would then be something like the following:

```rust
use futures::executor::ThreadPool;
let pool = ThreadPool::new().unwrap();

let i1 = pool.spawn_ok(render(icon_1));
let i2 = pool.spawn_ok(render(icon_2));
let i3 = pool.spawn_ok(render(icon_3));
let h2 = pool.spawn_ok(async{join!(i1, i2, i3); render(horizontal_2)});
let tp = pool.spawn_ok(async{join!(h2); render(top_panel)});

let cp = pool.spawn_ok(render(center_panel));

let t1 = pool.spawn_ok(render(tool_1));
let t2 = pool.spawn_ok(render(tool_2));
let t3 = pool.spawn_ok(render(tool_3));
let h1 = pool.spawn_ok(async{join!(t1, t2, t3); render(horizontal_1)});
let bp = pool.spawn_ok(async{join!(h1); render(bottom_panel)});

let win = pool.spawn_ok(async{join!(tp, cp, bp); render(Window)});

let gui = pool.spawn_ok(async{join!(win); render(win).await});
```
(I probably scrambled some of my `async` stuff, but that should give you an idea of what I was thinking of).

Basically, I've given egui a bunch of work, and specified that for containers to be laid out, their contents need to be laid out first.  This is **not** retained mode!  You have to do this each and every time you want to paint the GUI.  

Note that this is probably a poor way to do it; ideally, you'd create the GUI instance first so that you can pass it down to its children (in case they need to know something about their parent container), spawn children, which spawn more children, etc., until you finally start hitting the leaves.
